File: bin/generate_indexes.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Generate navigation index.md files for the mechanically-extracted
# `references/` and `style-guide/` trees.
#
# These trees are NOT part of the OKF bundle (their files have no OKF
# frontmatter); the generated index.md files are plain-Markdown navigation
# aids, kept current by `make references` / `make style-guide`.
#
# Usage: bin/generate_indexes.rb [references|style-guide|all]

require "pathname"

REPO = Pathname.new(__dir__).parent

AREAS = %w[rails hotwire-native stimulus strada turbo kamal].freeze
AREA_TITLES = {
  "rails" => "Rails Guides", "hotwire-native" => "Hotwire Native",
  "stimulus" => "Stimulus", "strada" => "Strada", "turbo" => "Turbo",
  "kamal" => "Kamal"
}.freeze

# Whole-word fixups for names that simple capitalization gets wrong.
ACRONYMS = {
  "rspec" => "RSpec", "css" => "CSS", "html" => "HTML", "api" => "API",
  "cli" => "CLI", "ssh" => "SSH", "ssl" => "SSL", "sql" => "SQL",
  "url" => "URL", "io" => "IO", "http" => "HTTP", "yaml" => "YAML",
  "json" => "JSON", "typescript" => "TypeScript"
}.freeze

def titleize(slug)
  slug.sub(/\A\d+[_-]/, "").gsub(/[_-]+/, " ").strip.split.map do |w|
    next w if w.empty?

    ACRONYMS[w.downcase] || (w[0].upcase + w[1..])
  end.join(" ")
end

def unquote(str)
  s = str.strip
  s = s[1..-2] if s.length >= 2 && %w[" '].include?(s[0]) && s[-1] == s[0]
  s
end

# First real title: frontmatter `title:`, else first ATX/Setext heading in the
# body (skipping the "DO NOT READ" / generated-from-source preamble), else a
# titleized filename.
def doc_title(path)
  lines = File.read(path).lines.map(&:chomp)
  body_start = 0
  if lines.first == "---"
    close = lines[1..].index("---")
    if close
      fm = lines[1..close]
      t = fm.find { |l| l =~ /\Atitle:/ }
      return unquote(t.split(":", 2)[1]) if t

      body_start = close + 2 # skip past the closing ---
    end
  end
  lines[body_start, 20].to_a.each_with_index do |l, i|
    s = l.strip
    next if s.empty? || s.start_with?("**", "<")
    return s[2..].strip if s.start_with?("# ")

    nxt = lines[body_start + i + 1]
    return s if nxt && nxt.strip.match?(/\A=+\z/) # Setext H1
  end
  titleize(File.basename(path, ".md"))
end

def md_files(dir)
  Pathname.glob(dir.join("**/*.md")).reject { |p| p.basename.to_s == "index.md" }.sort
end

def gen_references
  base = REPO.join("references")
  out = +"# References\n\n"
  out << "> Documentation mechanically extracted from upstream git submodules " \
         "(Rails guides, Hotwire, Kamal, RuboCop) via `make references`. These " \
         "files are **not** part of the OKF bundle — they carry no OKF frontmatter. " \
         "Regenerate this index with `make references-index`.\n"

  AREAS.each do |area|
    adir = base.join(area)
    next unless adir.directory?

    files = md_files(adir)
    next if files.empty?

    # Prefix each entry with its immediate-parent subsection only when the area
    # spans more than one subsection (e.g. Kamal's configuration/commands/hooks).
    subsections = files.map { |f| f.parent.basename.to_s }.uniq
    prefix = subsections.length > 1

    label = lambda do |f, text|
      prefix ? "#{titleize(f.parent.basename.to_s)}: #{text}" : text
    end
    entries = files.map { |f| [f, label.call(f, doc_title(f))] }

    # Disambiguate duplicate labels (e.g. Kamal's docs.md and help.md both
    # carry `title: Help` upstream) by falling back to the titleized filename.
    dup = entries.map { |_, t| t }.tally
    entries = entries.map do |f, t|
      dup[t] > 1 ? [f, label.call(f, titleize(File.basename(f, ".md")))] : [f, t]
    end

    out << "\n## #{AREA_TITLES.fetch(area, titleize(area))}\n\n"
    entries.each { |f, t| out << "* [#{t}](#{f.relative_path_from(base)})\n" }
  end
  write(base.join("index.md"), out)
end

def gen_style_guide
  base = REPO.join("style-guide")
  out = +"# Style Guides\n\n"
  out << "> RuboCop community style guides extracted from upstream submodules via " \
         "`make style-guide`. Not part of the OKF bundle. Regenerate this index " \
         "with `make style-guide-index`.\n\n"
  Dir.children(base).select { |f| f.end_with?(".adoc") }.sort.each do |f|
    out << "* [#{titleize(File.basename(f, '.adoc'))} Style Guide](#{f})\n"
  end
  write(base.join("index.md"), out)
end

def write(path, text)
  File.write(path, text.gsub(/\n{3,}/, "\n\n"))
  puts "wrote #{Pathname.new(path).relative_path_from(REPO)}"
end

what = ARGV[0] || "all"
gen_references if %w[references all].include?(what)
gen_style_guide if %w[style-guide all].include?(what)

File: bin/generate_indexes.py
#!/usr/bin/env python3
"""Generate navigation index.md files for the mechanically-extracted
`references/` and `style-guide/` trees.

These trees are NOT part of the OKF bundle (their files have no OKF
frontmatter); the generated index.md files are plain Markdown navigation
aids, kept current by `make references` / `make style-guide`.

Usage: generate_indexes.py [references|style-guide|all]
"""
import os, re, sys

REPO = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))

AREA_TITLES = {
    "rails": "Rails Guides",
    "hotwire-native": "Hotwire Native",
    "stimulus": "Stimulus",
    "strada": "Strada",
    "turbo": "Turbo",
    "kamal": "Kamal",
}
SUBDIR_TITLES = {
    "guides": "Guides", "overview": "Overview", "reference": "Reference",
    "handbook": "Handbook", "docs": "Docs", "commands": "Commands",
    "configuration": "Configuration", "hooks": "Hooks",
    "installation": "Installation", "upgrading": "Upgrading", "search": "Search",
}

def titleize(slug):
    return re.sub(r"[_-]+", " ", slug).strip().title()

def doc_title(path):
    """First real heading: skip the 'DO NOT READ' preamble; handle Setext (===)
    and ATX (# ) headings; fall back to a titleized filename."""
    try:
        lines = open(path, encoding="utf-8").read().split("\n")
    except OSError:
        return titleize(os.path.basename(path)[:-3])
    for i, l in enumerate(lines[:15]):
        s = l.strip()
        if not s or s.startswith("**") or s.startswith("<"):
            continue
        if s.startswith("# "):
            return s[2:].strip()
        if i + 1 < len(lines) and re.fullmatch(r"=+", lines[i + 1].strip()) and s:
            return s  # Setext H1
    return titleize(os.path.basename(path)[:-3])

def walk_md(root):
    out = []
    for dirpath, _, names in os.walk(root):
        for n in sorted(names):
            if n.endswith(".md") and n != "index.md":
                out.append(os.path.join(dirpath, n))
    return sorted(out)

def gen_references():
    base = os.path.join(REPO, "references")
    lines = [
        "# References", "",
        "> Documentation mechanically extracted from upstream git submodules "
        "(Rails guides, Hotwire, Kamal, RuboCop) via `make references`. These "
        "files are **not** part of the OKF bundle — they carry no OKF frontmatter. "
        "Regenerate this index with `make references-index`.", "",
    ]
    for area in ["rails", "hotwire-native", "stimulus", "strada", "turbo", "kamal"]:
        adir = os.path.join(base, area)
        if not os.path.isdir(adir):
            continue
        files = walk_md(adir)
        if not files:
            continue
        lines += [f"## {AREA_TITLES.get(area, titleize(area))}", ""]
        cur_sub = None
        for f in files:
            rel = os.path.relpath(f, base)                      # e.g. rails/guides/x.md
            parts = rel.split(os.sep)
            sub = parts[1] if len(parts) > 2 else None          # subdir under area
            if sub != cur_sub:
                cur_sub = sub
                if sub:
                    lines += ["", f"### {SUBDIR_TITLES.get(sub, titleize(sub))}", ""]
            lines.append(f"* [{doc_title(f)}]({rel})")
        lines.append("")
    write(os.path.join(base, "index.md"), "\n".join(lines).rstrip() + "\n")

def gen_style_guide():
    base = os.path.join(REPO, "style-guide")
    files = sorted(f for f in os.listdir(base) if f.endswith(".adoc"))
    lines = [
        "# Style Guides", "",
        "> RuboCop community style guides extracted from upstream submodules via "
        "`make style-guide`. Not part of the OKF bundle. Regenerate this index "
        "with `make style-guide-index`.", "",
    ]
    for f in files:
        lines.append(f"* [{titleize(f[:-5])} Style Guide]({f})")
    write(os.path.join(base, "index.md"), "\n".join(lines) + "\n")

def write(path, text):
    text = re.sub(r"\n{3,}", "\n\n", text)   # collapse extra blank lines
    open(path, "w", encoding="utf-8").write(text)
    print("wrote", os.path.relpath(path, REPO))

if __name__ == "__main__":
    what = sys.argv[1] if len(sys.argv) > 1 else "all"
    if what in ("references", "all"):
        gen_references()
    if what in ("style-guide", "all"):
        gen_style_guide()
